src/einsteinpy/symbolic/tensor.py

```diff
@@ -1,55 +1,327 @@
+        >>> schwarzschild = diag(1-1/r, -1/(1-1/r), -r**2, -r**2*sin(th)**2)
+        >>> g = Metric('g', [t, r, th, ph], schwarzschild)
+        >>> mu, nu = indices('mu nu', g)
+        >>> g.covariance_transform(mu, nu)
+        [[1/(1 - 1/r), 0, 0, 0], [0, 1 - 1/r, 0, 0], [0, 0, -1/r**2, 0], [0, 0, 0, -1/(r**2*sin(theta)**2)]]
 
-        Returns
-        -------
-        ~sympy.tensor.array.dense_ndim_array.ImmutableDenseNDimArray
-            Sympy Array object
-        
         """
-        return self.arr
+        array = self.as_array()
+        for pos, idx in enumerate(indices):
+            if idx.is_up ^ (self.covar[pos] > 0):
+                if idx.is_up:
+                    metric = idx.tensor_index_type.metric.as_inverse()
+                else:
+                    metric = idx.tensor_index_type.metric.as_array()
+                new = tensorcontraction(tensorproduct(metric, array), (1, 2 + pos))
+                permu = list(range(len(indices)))
+                permu[0], permu[pos] = permu[pos], permu[0]
+                array = permutedims(new, permu)
+        return array
 
     def simplify(self):
         """
-        Returns a simplified Tensor
-
-        Returns
-        -------
-        ~einsteinpy.symbolic.tensor.Tensor
-            Simplified Tensor
+        Replace the stored array associated with this tensor with a simplified
+        version. This method also replaces the entry in ReplacementManager.
 
         """
-        return sympy.simplify(self.tensor())
+        array = simplify(self.as_array())
+        self._array = array
+        ReplacementManager.replace(self, array)
+        return array
+
+
+class Index(TensorIndex):
+    """
+    Class for a symbolic representation of a tensor index with respect to a metric.
+    """
+
+    def __new__(cls, symbol, metric, is_up=True, **kwargs):
+        return super().__new__(cls, symbol, metric, is_up=is_up, **kwargs)
+
+    def __neg__(self):
+        return Index(self.name, self.tensor_index_type, (not self.is_up))
+
+
+def expand_tensor(expr, idxs=None):
+    """
+    Evaluate a tensor expression and return the resulting array.
+
+    Parameters
+    ----------
+    expr : TensExpr
+        Symbolic expression of tensors.
+    idxs : TensorIndex
+        Indices that encode the covariance and contravariance of the result.
+
+    """
+    if idxs is None:
+        idxs = TensMul(expr).get_free_indices()
+    return expr.replace_with_arrays(ReplacementManager, idxs)
+
+
+def indices(s, metric, is_up=True):
+    """
+    Create indices using a method similar to ~sympy.symbols.
+    """
+    if isinstance(s, string_types):
+        a = [x.name for x in symbols(s, seq=True)]
+    else:
+        raise ValueError(f"expected a string, received object of type {type(s)}")
+    idxs = [Index(idx, metric, is_up) for idx in a]
+    if len(idxs) == 1:
+        return idxs[0]
+    return idxs
+
+
+# metric tensors and general tensors commute with each other and themselves.
```

Comment:
It's probably redundant to have these two commutation groups be separated. I'll probably remove the ``metric`` group.